### ISSUE TITLE
Website is down

### ISSUE THREAD

**Issue by turion** [haskell/criterion#278]
The website for criterion is linked in many places as http://www.serpentine.com/criterion, but it is offline.

**Comment by Lysxia** [comment]
Right now it looks online to me, but it's an http link, which may be a problem depending on browser extensions; and the https certificate is out of date.

**Comment by turion** [comment]
https://www.isitdownrightnow.com/serpentine.com.html says that the site is certainly down. Maybe you still have it cached?

**Comment by turion** [comment]
Also, the https variant doesn't work for me either.

**Comment by RyanGlScott** [comment]
I am also able to view the website, and I don't have it cached.

That being said, I think it is time to consider migrating the website away from www.serpentine.com to a different place. (Perhaps somewhere on github.io using GitHub pages?) I don't own www.serpentine.com, which means that there's nothing I could do if the website truly goes down. Nor can I update the website should we change the material presented on that page, as was recently done in https://github.com/haskell/criterion/pull/268.

**Comment by turion** [comment]
When I SSH into my webserver, I can curl the site (but only http). From my personal laptop, I cannot curl it. It says:

```
❯ curl http://www.serpentine.com/criterion/tutorial.html
curl: (6) Could not resolve host: www.serpentine.com
```

So there is probably something wrong with DNS, but not only on my laptop, since isitdownrightnow.com says it's down as well. I checked two other similar services: https://www.checkifdown.com/ sometimes says it is _up_ (although a 500 error), sometimes it is down, https://downforeveryoneorjustme.com/serpentine.com?proto=http&www=1 says it is up.

Maybe a cloudflare issue?

**Comment by turion** [comment]
github.io or readthedocs sound like sensible places.

**Comment by RyanGlScott** [comment]
If my understanding of how GitHub Pages works is correct, we'd need to:

1. Enable the appropriate options in https://github.com/haskell/criterion/settings/pages. (I'd need a @haskell administrator to help me with this step, as I don't have the permissions to do this.)
2. Add a GitHub Actions workflow that uses the [`deploy-pages`](https://github.com/actions/deploy-pages) action to build the website (i.e., navigate to the [`www`](https://github.com/haskell/criterion/tree/b6713b7f343ce4ef9f097608f6238bec365d5666/www) subdirectory and run `make`) and deploy it.

**Comment by ulysses4ever** [comment]
The website is down again. It'd be great to fix this issue, e.g. as discussed above, via GitHub Pages. Should we scream about the insufficient permissions (to enable Pages) at https://github.com/haskell/meta/ ? I am happy to assist if everyone is busy. (I'm not an admin but I know how to scream.)

I feel like sincce the tutorial is just one page, maybe it should simply be included in sdist?.. After all, people include whole mdbooks in Hackage releases: https://hackage.haskell.org/package/text-display/docs/doc/book/Introduction.html Making sure its packaging is correctly automated may be more hussle then github pages though.

**Comment by Bodigrim** [comment]
Why not shove it all into `README.md`?

**Comment by ulysses4ever** [comment]
Yes, that's an option. I seem to remember that it's hard to have image paths work properly on both Hackage and GitHub, but maybe with enough care it's feasible?..

<bikeshedding>I kinda like the current readme, so I'd maybe create a `TUTORIAL.md` or something.</bikeshedding> But the idea is the same: just convert the current tutorial from HTML to Markdown, and then let GitHub do its thing...

**Comment by Bodigrim** [comment]
> Yes, that's an option. I seem to remember that it's hard to have image paths work properly on both Hackage and GitHub, but maybe with enough care it's feasible?..

The trick is to link an image from Markdown by its absolute URL, hosted at Hackage. And add an image to `extra-doc-files`. You can check how `tasty-bench` embeds `example.svg`: https://github.com/Bodigrim/tasty-bench/blob/master/README.md#plotting-results and https://hackage.haskell.org/package/tasty-bench-0.4.1#plotting-results.

**Comment by RyanGlScott** [comment]
I don't have a strong opinion about whether we upload the tutorial to GitHub Pages or include it in the `README`. The latter would have the advantage of being something that we could do now (without needing a haskell/meta administrator to intervene).

**Comment by Bodigrim** [comment]
Regarding GitHub Pages, we don't have anything at `haskell.github.io` right now and it does not feel right to squat it solely for Criterion documentation. So, I assume, one needs a custom domain, paid and maintained. 

Putting entire documentation into `README.md` is much more robust approach: even if GitHub and Hackage burst in flames, the documentation is still right there, in every tarball on every machine.

**Comment by ulysses4ever** [comment]
@Bodigrim the whole domain wouldn't be squated by criterion solely: any project in this org can use haskell.github.io/${project} for its GitHub pages (see the table under https://docs.github.com/en/pages/getting-started-with-github-pages/what-is-github-pages#types-of-github-pages-sites, the Project Sites / Default site location). But I completely agree that for one simple page it's an overkill.

**Comment by ulysses4ever** [comment]
I wasted some time on cleaning up the output of Pandoc-conversion HTML→MD only to find out that there's already a markdown version: https://github.com/haskell/criterion/blob/master/www/tutorial.md...

A little issue with outsorcing the MD rendering is that the Tutorial references an example report, which is, obviously, an HTML file. I don't see a good way to reference it in a way that would allot to open that report easily with one click as it was the case with the original tutorial...

**Comment by RyanGlScott** [comment]
Nice, thanks for looking into this! Some follow-up questions:

* You mentioned needing to clean up the output of `pandoc`... is this something that can be automated, or is it non-trivial to do so?
* Do you think we could set things up such that we could generate the entire `README.md` (both the current contents as well as the tutorial) as part of the `pandoc`-related scripting?

**Comment by ulysses4ever** [comment]
I think both can be done, but my question in return: why would you want to preserve the HTML (to use it as the input for the MD-generation mechanism)? I'd migrate it to MD and forget the HTML. 

The page is pretty simple so it's a good target for MD. The only question I had is for the little fancier elements (namely, tables and callouts), we'll need to decide what to do (if we decide to migrate to MD).

**Comment by ulysses4ever** [comment]
> preserve the ability to generate an HTML version of the tutorial

that should be trivial, with no postprocessing steps. It's the other way round (HTML→MD) that is a tad more painful.

>  Using the standard GitHub-Flavored Markdown conventions whereever appropriate (e.g., using GFM's syntax for tables) would be preferable.

Great, thanks for deciding on that! As the current maintainer, you will have to make decisions! :-) Here are two more questions.

GFM is only somewhat standard. For instance, I doubt Hackage will render GFM tables (haven't checked, though). So, if we converge on GFM (which I think is a good idea), we'll have to have a separate .md file and reference it from the README. Does this sound right? If We can include that .md file in the tarball just in case of an end-of-civilization event as portrayed above by Bodigrim.

The last question I have is about the sample HTML report that is referenced from the tutorial: are we okay with it not being readily available for rendering? Basically, if we use GitHub as our MD rendering engine, then to browse the HTML report, the user will have to download the HTML file and open it in the browser separately (if I'm not mistaken). This is a step down in the quality of life from when we had dedicated hosting. Especially given that the tutorial touts the report's interactive features. If we want the report easily available, we'd need to go the GitHub Pages route. What's your stance on that @RyanGlScott? I'm on the fence about this one: I like the simplicity of GitHub-rendered MD, but I'll be sad to see the abovementioned degradation in life quality.

**Comment by Bodigrim** [comment]
FWIW Hackage copes reasonably well with rendering GFM tables, see https://hackage.haskell.org/package/mod.

**Comment by ulysses4ever** [comment]
That's cool! Do you know about callouts? (https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts) Do you have an opinion about the HTML report and the fact that if we don't have a server for it (e.g. GitHub Pages), people won't be able to access it as easily?

**Comment by Bodigrim** [comment]
> Do you know about callouts?

See https://github.com/haskell/hackage-server/blob/1d5f29178aa0eb6bacf6a72764b6d3ba158c76d0/src/Distribution/Server/Util/Markdown.hs#L200-L213
[`alertSpec`](https://hackage-content.haskell.org/package/commonmark-extensions-0.2.6/docs/Commonmark-Extensions-Alerts.html#v:alertSpec) is not enabled at the moment, but I think it's simply because no one asked about it. 

> The last question I have is about the sample HTML report that is referenced from the tutorial: are we okay with it not being readily available for rendering? Basically, if we use GitHub as our MD rendering engine, then to browse the HTML report, the user will have to download the HTML file and open it in the browser separately (if I'm not mistaken).

I'm not sure I understand the problem correctly. Are we talking about [`fibber.html`](https://github.com/haskell/criterion/blob/master/www/fibber.html)? Why cannot we put it inside the package under `extra-doc-files` and link it on Hackage?

**Comment by RyanGlScott** [comment]
Generally speaking, I think it would be nice to include as much of the tutorial in the README as possible, but I do acknowledge the fact that we won't be able to put everything in it (e.g., the HTML report). My inclination (without having looked too much into the implementation specifics) is to include the HTML report from `extra-doc-files` and then find a reasonable way to link to the report from the README. I think that should strike a reasonable balance of discoverability while still including everything one would want to read in a `criterion` sdist.

**Comment by ulysses4ever** [comment]
Sounds good, thanks everybody. I'll try to submit a PR by the end of the week

**Comment by RyanGlScott** [comment]
I've merged #284 and released these changes to Hackage as [`criterion-1.6.4.1`](https://hackage.haskell.org/package/criterion-1.6.4.1). With this, the tutorial is now fully included in the `criterion` README, which should be viewable on GitHub, Hackage, or in locally downloaded sdists.

**Comment by ulysses4ever** [comment]
GitHub tag line should probably be updated as well. Hackage page seems like a good target.

<img width="864" height="1920" alt="Image" src="https://github.com/user-attachments/assets/75a05699-3c9a-43d0-8fff-6ac50f413360" />

**Comment by RyanGlScott** [comment]
Unfortunately, I don't appear to have the permissions needed to edit the tag line.

**Comment by ulysses4ever** [comment]
Hmm... This sounds like something that needs fixing. Should we reach to the haskell org admins (via haskell/meta)? Google suggests that they should be able to add admins to individual repos.

**Comment by RyanGlScott** [comment]
Yes, we should definitely reach out to an @haskell admin. I'm not actually sure who the admins are, however...

**Comment by RyanGlScott** [comment]
Ah, I wasn't aware of https://github.com/haskell/meta! When I get a chance, I'll open an issue there.

**Comment by RyanGlScott** [comment]
https://github.com/haskell/meta/issues/10
